core/src/main/scala/com/rallyhealth/vapors/core/data/ResultSet.scala

```diff
@@ -1,84 +1,122 @@
 import cats.data.NonEmptyList
+import com.rallyhealth.vapors.core.dsl.{AbstractFacts, Facts}
 import com.rallyhealth.vapors.core.logic.{Intersect, Union}
 
-sealed trait ResultSet[+A] {
+sealed abstract class AbstractResultSet {
+  type FactValue
 
-  def matchingFacts: List[Fact[A]]
+  def matchingFacts: List[AbstractFact]
 
   /**
     * Union of the two sets of results.
     */
-  def union[B >: A](o: ResultSet[B]): ResultSet[B]
+  def union(o: AbstractResultSet): AbstractResultSet
 
   /**
     * Alias for [[union]].
     */
-  def ++[B >: A](o: ResultSet[B]): ResultSet[B]
-}
-
-object ResultSet {
-
-  def fromNel[I](nonEmptyFacts: NonEmptyList[Fact[I]]): FactsMatch[I] = FactsMatch(nonEmptyFacts)
+  @inline final def ++(o: AbstractResultSet): AbstractResultSet = this.union(o)
 
-  def fromList[I](facts: List[Fact[I]]): ResultSet[I] = NonEmptyList.fromList(facts) match {
-    case Some(nonEmptyList) => FactsMatch(nonEmptyList)
-    case None => NoFactsMatch()
+  override final def equals(o: Any): Boolean = o match {
+    case that: AbstractResultSet =>
+      (this eq that) || this.matchingFacts == this.matchingFacts
+    case _ => false
   }
 
-  // TODO: Should this be a MonoidK? It's feasible now, but maybe it isn't what we want
-  implicit object Instances extends SemigroupK[ResultSet] {
-    override def combineK[A](
-      x: ResultSet[A],
-      y: ResultSet[A],
-    ): ResultSet[A] = x ++ y
+  override final def toString: String = this match {
+    case NoFactsMatch() => "NoFactsMatch()"
+    case matched: AbstractFactsMatch => s"FactsMatch(${matched.matchingFactsNel})"
   }
+}
+
+object AbstractResultSet {
 
-  implicit def unionAny[T]: Union[ResultSet[T]] = {
-    _.reduceLeft[ResultSet[T]] {
+  implicit val unionAbstractResultSet: Union[AbstractResultSet] = {
+    _.reduceLeft[AbstractResultSet] {
       case (NoFactsMatch(), b) => b // try the next expression if the first failed
       case (a, NoFactsMatch()) => a // use the previous expression if the next fails
       case (a, b) => a ++ b // union all the possible facts (for better quality calculations)
     }
   }
 
-  implicit def intersectAny[T]: Intersect[ResultSet[T]] = {
-    _.reduceLeft[ResultSet[T]] {
+  implicit val intersectAbstractResultSet: Intersect[AbstractResultSet] = {
+    _.reduceLeft[AbstractResultSet] {
       case (NoFactsMatch(), _) | (_, NoFactsMatch()) => NoFactsMatch() // skip the all expressions if the first failed
       case (acc, nextResult) => acc ++ nextResult // union all the required facts
     }
   }
 }
 
-final case class FactsMatch[+I](matchingFactsNel: NonEmptyList[Fact[I]]) extends ResultSet[I] {
+sealed trait ResultSet[A] extends AbstractResultSet {
+  override type FactValue = A
 
-  override def matchingFacts: List[Fact[I]] = matchingFactsNel.toList
+  override def matchingFacts: List[Fact[A]]
+}
+
+object ResultSet {
+
+  def fromNel[I](nonEmptyFacts: NonEmptyList[Fact[I]]): FactsMatch[I] = FactsMatch(nonEmptyFacts)
```

Comment:
Unify to use 'A' as the type parameter?

Reply:
Good call. This is inconsistent with elsewhere. I've fixed it.